crate2nix/templates/nix/crate2nix/default.nix

```diff
@@ -571,6 +573,40 @@ rec {
     then builtins.throw "strictDeprecation enabled, aborting: ${message}"
     else builtins.trace message value;
 
+  /* create a shell express for the given crates from the crateOverrides */
+  buildWorkspaceShell = crateOverrides: crates:
+    assert builtins.isAttrs crates;
+    assert builtins.isAttrs crateOverrides;
+    let
+      overrides =
+        lib.attrValues
```

Comment:
I agree that it is only a first step. We will have to look into a proper way of collecting the overrides. My main motivation was to reduce the amount of eval time spent on this. The very first version I wrote was recrusively going through all the `buildRustCrate` dependencies and did collect from them. That was very expensive as (most) of the dependencies occur plentiful throughout the entire tree. Just looking at the list of defined crates made this >4x faster and actually more useable. We also do not have to instantiate a single `buildRustCrate` derivation that way.

That all being said, you are totally right that we might want to use a similar approach to what you pointed out. I'll have to look into that in more detail to understand how to apply it in this scenario.

Reply:
To avoid the recursion and processing dependencies double, you can also

1. filter the crates by the target if there is an OS restriction (that's the most important),
2. at least parse the crateName and version to the override function.

Reply:
I thought about this a bit more. Filtering by feature is probably not desirable. I would expect the shell expression to be able to build all the features of my crates. Filtering by OS is another story and that should probably be implemented. I'll revisit the implementation again.

Reply:
Thanks for taking the time to think about this!

About features: I think that if you want to be precise (which is not super important) and still make sure that you include everything relevant, you'd still filter by feature but assume all of them to be enabled for the root project.

If you, in addition, want to avoid that users have to reopen the shell when they add new features, you could enable all features for direct dependencies instead.

Features can enable a lot of unneeded stuff but that said, the harm in having too much in the environment is much smaller than not enough. Except you hid something that doesn't work and that you won't need anyways...